src/org/zaproxy/zap/extension/csphelper/CspHelper.java

```diff
@@ -0,0 +1,467 @@
+				csp.getScriptSource().addUrl(src);
+			} else {
+				// Handle inline scripts
+				String content = script.getContent().toString();
+				csp.getScriptSource().addHashNonce("'" + 
+					ExtensionCspHelper.sha256(content) +
+					"'");
+				logger.debug("SBSB adding hash " + ExtensionCspHelper.sha256(script.getContent().toString())); // TODO
+				handleWorkerSources(content, csp);
+				handleConnectSources(content, csp);
+			}
+		}
+		logger.debug("SBSB csp now " + csp.getSite() + " : " + csp.generate()); // TODO
+	}
+	
+	private void handleWorkerSources(String content, CSP csp) {
+		// Regex checks to identify connect-src, worker-src etc
+		ArrayList<String> workers = regexGetGroup("(Worker|\\.register)\\([\"']([\\w\\.:/?=&]*)[\"']", 2, content);
+		for (String w : workers) {
+			csp.getWorkerSource().addUrl(w);
+		}
+	}
+	
+	private void handleConnectSources(String content, CSP csp) {
+		// Regex checks to identify connect-src
+		// (WebSocket|EventSource|\\.sendBeacon)\\([\"']([\\w\\.:/?=&#]*)[\"']
+		// \\.open\\([\"'][a-zA-Z]+[\"'],\\s*[\"']([\\w\\.:/?=&#]*)[\"']
+		ArrayList<String> connects = regexGetGroup("[\"']((http[s]?:\\/\\/)?\\w+\\.\\w[\\.\\w/?=&]+)[\"']", 1, content);
+		for (String c : connects) {
+			switch(getLinkType(c)) {
+				case LINK_TYPE_IMAGE:
+					csp.getImageSource().addUrl(c);
+					break;
+				case LINK_TYPE_STYLE:
+					csp.getStyleSource().addUrl(c);
+					break;
+				case LINK_TYPE_SCRIPT:
+					csp.getScriptSource().addUrl(c);
+					break;
+				case LINK_TYPE_REQUEST:
+					csp.getConnectSource().addUrl(c);
+					break;
+				default:
+					break;
+			}
+		}
+	}
+	
+	private void handleImages(CSP csp, HttpMessage msg, Source source) {
+		// Images tags
+		List<Element> images = source.getAllElements("img");
+		for (Element image : images) {
+			String src = image.getAttributeValue("src");
+			if (src != null) {
+				csp.getImageSource().addUrl(src);
+			}
+		}
+	}
+	
+	private void handleMedia(CSP csp, HttpMessage msg, Source source) {
+		// Audio tags
+		List<Element> audio = source.getAllElements("audio");
+		for (Element a : audio) {
+			List<Element> inner = a.getChildElements();
+			for (Element childEl : inner) {
+				String src = childEl.getAttributeValue("src");
+				if (src != null) {
+					csp.getMediaSource().addUrl(src);
+				}
+			}
+		}
+		// Audio tags
+		List<Element> video = source.getAllElements("video");
+		for (Element v : video) {
+			List<Element> inner = v.getChildElements();
+			for (Element childEl : inner) {
+				String src = childEl.getAttributeValue("src");
+				if (src != null) {
+					csp.getMediaSource().addUrl(src);
+				}
+			}
+		}
+	}
+	
+	private void handleFrames(CSP csp, HttpMessage msg, Source source) {
+		// Frameset tags
+		List<Element> framesets = source.getAllElements("frameset");
+		for (Element frame : framesets) {
+			List<Element> inner = frame.getChildElements();
+			for (Element childEl : inner) {
+				String src = childEl.getAttributeValue("src");
```

Comment:
This and following statements could be extracted to a method, same code as for the iframes.